data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,173 @@
+        this.recordProcessingErrors = pluginMetrics.counterWithTags(KINESIS_RECORD_PROCESSING_ERRORS, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointFailures = pluginMetrics.counterWithTags(KINESIS_CHECKPOINT_FAILURES, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointInterval = kinesisStreamConfig.getCheckPointInterval();
+        this.buffer = buffer;
+    }
+
+    private KinesisStreamConfig getStreamConfig(final KinesisSourceConfig kinesisSourceConfig) {
+        return kinesisSourceConfig.getStreams().stream().filter(streamConfig -> streamConfig.getName().equals(streamIdentifier.streamName())).findAny().get();
+    }
+
+    @Override
+    public void initialize(InitializationInput initializationInput) {
+        // Called once when the processor is initialized.
+        kinesisShardId = initializationInput.shardId();
+        LOG.info("Initialize Processor for shard: {}", kinesisShardId);
+        lastCheckpointTimeInMillis = System.currentTimeMillis();
+    }
+
+    private AcknowledgementSet createAcknowledgmentSet(final ProcessRecordsInput processRecordsInput) {
+        return acknowledgementSetManager.create((result) -> {
+            acknowledgementSetCallbackCounter.increment();
+            if (result) {
+                LOG.info("acknowledgements received");
+                checkpoint(processRecordsInput.checkpointer());
+            } else {
+                LOG.info("acknowledgements received with false");
+            }
+
+        }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+        List<Record<Event>> records = new ArrayList<>();
+
+        try {
+            Optional<AcknowledgementSet> acknowledgementSetOpt = Optional.empty();
+            boolean acknowledgementsEnabled = kinesisSourceConfig.isAcknowledgments();
+            if (acknowledgementsEnabled) {
+                acknowledgementSetOpt = Optional.of(createAcknowledgmentSet(processRecordsInput));
+            }
+
+            for (KinesisClientRecord record : processRecordsInput.records()) {
+                processRecord(record, records::add);
+            }
+
+            acknowledgementSetOpt.ifPresent(acknowledgementSet -> records.forEach(record -> acknowledgementSet.add(record.getData())));
+
+            buffer.writeAll(records, bufferTimeoutMillis);
+
+            acknowledgementSetOpt.ifPresent(AcknowledgementSet::complete);
+
+            // Checkpoint for shard
+            if (kinesisStreamConfig.isEnableCheckPoint() && System.currentTimeMillis() - lastCheckpointTimeInMillis > checkpointInterval.toMillis()) {
+                LOG.info("Regular checkpointing for shard " + kinesisShardId);
+                checkpoint(processRecordsInput.checkpointer());
```

Comment:
I believe we should not call this when end-to-end acknowledgements are enabled. Otherwise, we will checkpoint prematurely.